docs/designs/endpoint-api.md

```diff
@@ -0,0 +1,164 @@
+# Endpoint API design
+
+## Principles
+
+The following principles should be followed when designing the endpoint API:
+
+- [ ] Prefer explicitness over implicitness
+- [ ] Make documentation a first-class citizen
+- [ ] Ensure type-safe decorators
+- [ ] Define endpoints in a consistent manner
+- [ ] Ensure the API is clean and easy to use/understand
+- [ ]
+
+## Endpoint definition
+
+Leads:
+
+- @jonnydgreen
+- @aleccool213
+
+### Controller
+
+```typescript
+@Controller({
+  route: "/hello",
+})
+export class Controller {}
+```
+
+### Router
+
+```typescript
+export class Controller {
+  @Post({
+    description: "Some description.",
+    route: "/there",
+    response: Hello,
+  })
+  public getHelloThere(): Hello {
+    return {
+      name: "name",
+      address: "address",
+    };
+  }
+}
+```
+
+## Type definitions
+
+A lot of issues with current frameworks is the inconsistency in defining types.
+This is a proposal to define types in a consistent manner with the use of
+type-safe decorators.
+
+### Basic
+
+Lead: @jonnydgreen
+
+```typescript
+@Object({
+  description: "Some description.",
+  examples: [{
+    name: "Hello",
+    address: "1 Strawberry Fields, Fruitsville, B12 YUM",
+  }],
+})
+export class Hello {
+  @Field({
+    description: "Some description.",
+    examples: ["Hello"],
+    type: String,
+  })
+  name!: string;
+
+  @Field({
+    description: "Some description.",
+    examples: ["1 Strawberry Fields, Fruitsville, B12 YUM"],
+    type: String,
+  })
+  address!: string;
+}
+```
+
+### List
+
+Lead: @jonnydgreen
+
+#### Usage
+
+```typescript
+export class HelloController {
+  @Post({ outputType: List(Hello) })
+  public getHelloList(): Hello[] {
```

Comment:
@aleccool213 Because of limitations in being able to define a `Tuple` type, I left this as the syntactically more common usage - unless you have any ideas regarding multiple arbitrary args to a tuple?